### PR TITLE
feedback modal, emails

### PR DESCRIPTION
**Pull request by rsimosa**
Adding Feedback modal and thankyou modals
Added a check for not sending any information to the backEnd if the form has no input
Added SendGrid to send feedback to an email
Added Unit Tests for the notification Manager

image
image

### CODE REVIEW

File: src/DashboardUI/src/components/SiteFooter.tsx [hunk 1]
```diff
@@ -21,9 +26,10 @@ function SiteFooter() {
         <div>
           <Nav>
             <Nav.Link className="p-0" target="_blank" rel="noopener noreferrer" href="https://westernstateswater.org/wade">FAQ</Nav.Link>
-            <Nav.Link className="p-0" target="_blank" rel="noopener noreferrer" href="#">Feedback</Nav.Link>
+            <Nav.Link className="p-0" target="_blank"  href="#" onClick={() => props.showFeedbackModal(true)}> Feedback </Nav.Link>
```
Comment (Rhipeus) [inline src/DashboardUI/src/components/SiteFooter.tsx:29]: This should be a button and not a link

**Review by Rhipeus — CHANGES_REQUESTED** [review]

File: src/API/WesternStatesWater.WestDaat.Managers/NotificationManager.cs [hunk 2]
```diff
@@ -0,0 +1,147 @@
...
+        {
+            _emailSDK = emailSDK;
+        }
+
+        async Task INotificationManager.SendFeedback(FeedbackRequest request)
+        {
+            var emailAddresses = new List<string> { "adelabdallah@wswc.utah.gov", "rjames@wswc.utah.gov" };
+            if (IsValidEmail(request.Email?.Trim()))
+            {
+                emailAddresses.Add(request.Email?.Trim());
```
[the remark this answers is not in the document]
Comment (rsimosa) [reply src/API/WesternStatesWater.WestDaat.Managers/NotificationManager.cs:24]: Adel required to send the email to the requestor
image

File: src/API/WesternStatesWater.WestDaat.Utilities/EmailNotificationSDK.cs [hunk 3]
```diff
@@ -0,0 +1,45 @@
...
+            foreach (var address in message.To)
+            {
+                msg.AddTo(address);
+            }
+
+            var response = await _client.SendEmailAsync(msg);
+
+            if (!response.IsSuccessStatusCode)
+            {
+                _logger.LogError($"Something went wrong while sending feedback email.\nStatusCode: {response.StatusCode}.\nMessageBody: {message.Body}.");
```
[the remark this answers is not in the document]
Comment (rsimosa) [reply src/API/WesternStatesWater.WestDaat.Utilities/EmailNotificationSDK.cs:39]: I did it this way to keep the message body and keep the feedback, as I only see it failing if we do more than the free tier allowed messages or SendGrid is down

File: src/API/WesternStatesWater.WestDaat.Managers/NotificationManager.cs [hunk 4]
```diff
@@ -0,0 +1,147 @@
...
+                    @"^[^@\s]+@[^@\s]+\.[^@\s]+$",
+                    RegexOptions.IgnoreCase, TimeSpan.FromMilliseconds(250));
+            }
+            catch
+            {
+                return false;
+            }
+        }
+
+        private string FeedbackMessageBody(FeedbackRequest request)
```
[the remark this answers is not in the document]
Comment (rsimosa) [reply src/API/WesternStatesWater.WestDaat.Managers/NotificationManager.cs:83]: sendGrid is ignoring the line breaks `\n` and it was displaying it as a single continuous string
image

that's why I was using a basic html format, if we are okay with using a continous single String I am okay with removing the html markup I added.

else I will update to use string replacements and decoding or create a regex whitelist characters

[the remark this answers is not in the document]
Comment (Rhipeus) [reply src/API/WesternStatesWater.WestDaat.Managers/NotificationManager.cs:83]: Were you doing `\r\n` or `\n` only?  It should be \r\n.

Also, according to random stackoverflow user, SendGrid has a field to specify text vs html (https://stackoverflow.com/questions/55852805/send-grid-allow-line-breaks-in-emails-with-plain-text).  If so, make sure it is set to text.

File: src/API/WesternStatesWater.WestDaat.Managers/NotificationManager.cs [hunk 5]
```diff
@@ -0,0 +1,147 @@
...
+        private readonly IEmailNotificationSDK _emailSDK;
+
+        public NotificationManager(IEmailNotificationSDK emailSDK, ILogger<NotificationManager> logger) : base(logger)
+        {
+            _emailSDK = emailSDK;
+        }
+
+        async Task INotificationManager.SendFeedback(FeedbackRequest request)
+        {
+            var emailAddresses = new List<string> { "adelabdallah@wswc.utah.gov", "rjames@wswc.utah.gov" };
```
[the remark this answers is not in the document]
Comment (rsimosa) [reply src/API/WesternStatesWater.WestDaat.Managers/NotificationManager.cs]: updated

File: src/API/WesternStatesWater.WestDaat.Common/Configuration/ConfigurationHelper.cs [hunk 6]
```diff
@@ -35,6 +39,7 @@ public static NldiConfiguration GetNldiConfiguration(this IConfiguration config)
     public static class ConfigurationRootNames
     {
         public const string Database = "Database";
+        public const string SendGrid = "SendGrid";
```
[the remark this answers is not in the document]
Comment (rsimosa) [reply src/API/WesternStatesWater.WestDaat.Common/Configuration/ConfigurationHelper.cs]: updated

File: src/API/WesternStatesWater.WestDaat.Managers/NotificationManager.cs (hunk 4 above)
[the remark this answers is not in the document]
Comment (rsimosa) [reply src/API/WesternStatesWater.WestDaat.Managers/NotificationManager.cs]: I was only using the `\n` adding the `\r\n` fixed it Thanks

File: src/API/WesternStatesWater.WestDaat.Managers/NotificationManager.cs (hunk 2 above)
[the remark this answers is not in the document]
Comment (rsimosa) [reply src/API/WesternStatesWater.WestDaat.Managers/NotificationManager.cs]: updated, from latest teams msg

File: src/API/WesternStatesWater.WestDaat.Tests.ManagerTests/NotificationManagerTests.cs [hunk 7]
```diff
@@ -0,0 +1,147 @@
...
+        [DataRow("almostvalid.com")]
+        [DataRow("testing@testing.com@com.com")]
+        [DataRow("not@valid@.com")]
+        [DataRow("@gmail.com")]
+        [DataRow("blank spaces@notValid.com")]
+        [DataRow("blankSpaces@not Valid.com")]
+        [DataRow("missing@notValid")]
+        [DataRow("")]
+        [DataRow(null)]
+        public async Task SendFeedback_DoesNotAddEmailToList(string invalidEmail)
```
[the remark this answers is not in the document]
Comment (rsimosa) [reply src/API/WesternStatesWater.WestDaat.Tests.ManagerTests/NotificationManagerTests.cs]: updated

File: src/DashboardUI/src/components/SiteFooter.tsx (hunk 1 above)
Comment (rsimosa) [reply src/DashboardUI/src/components/SiteFooter.tsx]: updated

File: src/API/WesternStatesWater.WestDaat.Client.Functions/Startup.cs [hunk 8]
```diff
@@ -36,7 +36,9 @@ public override void Configure(IFunctionsHostBuilder builder)
 
             builder.Services.AddScoped(a => configuration.GetDatabaseConfiguration());
             builder.Services.AddScoped(a => configuration.GetNldiConfiguration());
+            builder.Services.AddScoped(a => configuration.GetSMTPConfiguration());
```
Comment (Rhipeus) [inline src/API/WesternStatesWater.WestDaat.Client.Functions/Startup.cs:39]: Acronyms 3 characters or longer should be camel-cased.  All the "SMTP"s should be "Smtp"

https://docs.microsoft.com/en-us/dotnet/standard/design-guidelines/capitalization-conventions

File: src/API/WesternStatesWater.WestDaat.Client.Functions/Startup.cs [hunk 9]
```diff
@@ -51,6 +53,7 @@ public override void Configure(IFunctionsHostBuilder builder)
             builder.Services.AddTransient<ITestAccessor, TestAccessor>();
             builder.Services.AddTransient<IWaterAllocationAccessor, WaterAllocationAccessor>();
 
+            builder.Services.AddTransient<IEmailNotificationSDK, EmailNotificationSDK>();
```
Comment (Rhipeus) [inline src/API/WesternStatesWater.WestDaat.Client.Functions/Startup.cs:56]: The casing would apply to SDK here also

File: src/API/WesternStatesWater.WestDaat.Utilities/EmailNotificationSDK.cs [hunk 10]
```diff
@@ -2,44 +2,63 @@
...
 
         public EmailNotificationSDK(EmailServiceConfiguration emailService, ILogger<EmailNotificationSDK> logger)
         {
-            _client = new SendGridClient(emailService.APIKey);
+            _emailConfig = emailService;
             _logger = logger;
         }
 
-        public async Task<bool> SendEmail(CommonDTO.EmailRequest message)
+        public async void SendFeedback(CommonDTO.EmailRequest message)
```
Comment (Rhipeus) [inline src/API/WesternStatesWater.WestDaat.Utilities/EmailNotificationSDK.cs:21]: We went too far with making this not return a success boolean.  The async/await task stuff was still appropriate and helpful.  We just shouldn't have been returning the boolean.

File: src/API/WesternStatesWater.WestDaat.Utilities/EmailNotificationSDK.cs [hunk 11]
```diff
@@ -2,44 +2,63 @@
...
-            var response = await _client.SendEmailAsync(msg);
+                var response = await client.SendEmailAsync(message);
 
-            if (!response.IsSuccessStatusCode)
+                if (!response.IsSuccessStatusCode)
+                {
+                    throw new WestDaatException($"Something went wrong while sending email.\nStatusCode: {response.StatusCode}.\nSubject: {message.Subject}");
+                }
+            }
+            catch (Exception ex)
```
Comment (Rhipeus) [inline src/API/WesternStatesWater.WestDaat.Utilities/EmailNotificationSDK.cs:58]: We shouldn't be catching.  Let this go up the stack.

File: src/API/WesternStatesWater.WestDaat.Client.Functions/System.cs [hunk 12]
```diff
@@ -88,18 +88,16 @@ public async Task<IActionResult> GetRiverBasinPolygonsByName([HttpTrigger(Author
...
+            string requestBody = string.Empty;
             using (StreamReader streamReader = new StreamReader(request.Body))
             {
                 requestBody = await streamReader.ReadToEndAsync();
             }
             var feedbackRequest = JsonConvert.DeserializeObject<FeedbackRequest>(requestBody);
 
-            await _notificationManager.SendFeedback(feedbackRequest);
-
-            return new OkResult();
```
Comment (Rhipeus) [inline src/API/WesternStatesWater.WestDaat.Client.Functions/System.cs]: We should still be returning an `OkResult` object here

File: src/DashboardUI/src/components/SiteFooter.tsx [hunk 13]
```diff
@@ -21,9 +26,10 @@ function SiteFooter() {
         <div>
           <Nav>
             <Nav.Link className="p-0" target="_blank" rel="noopener noreferrer" href="https://westernstateswater.org/wade">FAQ</Nav.Link>
-            <Nav.Link className="p-0" target="_blank" rel="noopener noreferrer" href="#">Feedback</Nav.Link>
+            <button type="button" className="p-0 btn btn-text-color" onClick={() => props.showFeedbackModal(true)}>  Feedback </button>
```
Comment (Rhipeus) [inline src/DashboardUI/src/components/SiteFooter.tsx:29]: Does this look like the link above?  I expected this to have a `btn-link`.  If it looks correct, no worries.

File: src/API/WesternStatesWater.WestDaat.Common/Configuration/EmailServiceConfiguration.cs [hunk 14]
```diff
@@ -3,5 +3,7 @@
     public class EmailServiceConfiguration
     {
         public string APIKey {get; set;}
```
Comment (Rhipeus) [inline src/API/WesternStatesWater.WestDaat.Common/Configuration/EmailServiceConfiguration.cs:5]: All of these properties are needed for this to work.  Yet I don't see any default values setup.  What is the plan for deployment into QA?  Are you manually setting these up?

File: src/API/WesternStatesWater.WestDaat.Contracts.Client/FeedbackRequest.cs [hunk 15]
```diff
@@ -2,6 +2,8 @@
 {
     public class FeedbackRequest
     {
+        public string Name { get; set; }
```
Comment (Rhipeus) [inline src/API/WesternStatesWater.WestDaat.Contracts.Client/FeedbackRequest.cs:5]: Should this be `FirstName`?

File: src/API/WesternStatesWater.WestDaat.Utilities/IEmailNotificationSDK.cs [hunk 16]
```diff
@@ -5,6 +5,6 @@ namespace WesternStatesWater.WestDaat.Utilities
 {
     public interface IEmailNotificationSDK
     {
-        Task<bool> SendEmail(CommonDTO.EmailRequest message);
+        void SendFeedback(CommonDTO.EmailRequest message);
```
Comment (Rhipeus) [inline src/API/WesternStatesWater.WestDaat.Utilities/IEmailNotificationSDK.cs:8]: `SendEmail` was better.  This method should be generic and not related to any specific message.

File: src/DashboardUI/src/components/FeedbackModal.tsx [hunk 17]
```diff
@@ -137,28 +160,29 @@ function FeedbackModal(props: FeedBackModalProps) {
             Please let us know your feedback about the Water Data Exchange Data (WaDE) Dashboard.
           </p>
           <div className="mb-3">
-            <label className="form-label fw-bolder">Comments(Optional)</label>
-            <input type="text" className="form-control" placeholder="Anything will help us improve" onChange={(e) => setCommentsValue(e.target.value ?? "")} value={commentsValue} />
+            <label className="form-label fw-bolder">Name(Optional)</label>
```
Comment (Rhipeus) [inline src/DashboardUI/src/components/FeedbackModal.tsx:163]: These too should probably be First Name

File: src/DashboardUI/src/components/FeedbackModal.tsx [hunk 18]
```diff
@@ -218,10 +242,8 @@ function FeedbackModal(props: FeedBackModalProps) {
             )}
             <div className="mt-3">
               <p>
-                Lorem Ipsum is simply dummy text of the printing and typesetting industry. Lorem Ipsum has been the industry's
-                when an unknown printer took a galley of type and scrambled it to make a type specimen book. It has survived
-                into electronic typesetting, remaining essentially unchanged. It was popularised in the 1960s with the release
-                passages, and more recently with desktop publishing software like Aldus PageMaker including versions of Lorem
+                Lorem Ipsum is simply dummy text of the printing and typesetting industry.
```
Comment (Rhipeus) [inline src/DashboardUI/src/components/FeedbackModal.tsx:245]: Is Adel aware he needs to provide text here?

**Review by Rhipeus — DISMISSED** [review]

**Review by EverettRhodes — APPROVED** [review]

**Pull request merged**